src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -699,37 +699,37 @@ private void ReturnReadBufferIfEmpty()
 
         private bool HaveFullTlsFrame(out int frameSize)
         {
-            if (_buffer.EncryptedLength < TlsFrameHelper.HeaderSize)
-            {
-                frameSize = int.MaxValue;
-                return false;
-            }
-
             frameSize = GetFrameSize(_buffer.EncryptedReadOnlySpan);
             return _buffer.EncryptedLength >= frameSize;
         }
 
         [AsyncMethodBuilder(typeof(PoolingAsyncValueTaskMethodBuilder<>))]
-        private async ValueTask<int> EnsureFullTlsFrameAsync<TIOAdapter>(CancellationToken cancellationToken)
+        private async ValueTask<int> EnsureFullTlsFrameAsync<TIOAdapter>(CancellationToken cancellationToken, int estimatedSize)
             where TIOAdapter : IReadWriteAdapter
         {
-            int frameSize;
-            if (HaveFullTlsFrame(out frameSize))
+            if (HaveFullTlsFrame(out int frameSize))
             {
                 return frameSize;
             }
 
-            if (frameSize != int.MaxValue)
+            // Do zero byte read to conserve resources:
+            // 1. We postpone allocation of read buffers. For example Kestrel may have idle connections without buffer until new request comes in.
+            // 2. The underlying stream may need to allocate GCHandle for async IO. If we postpone that until we have some data, we may be able to avoid it.
```

Comment:
```suggestion
            // 2. The underlying stream may need to allocate GCHandle for async IO (on Windows). If we postpone that until we have some data, we may be able to avoid it.
```